packages/derive/src/lib.rs

```diff
@@ -0,0 +1,88 @@
+#[macro_use]
+extern crate syn;
+
+use proc_macro::TokenStream;
+use std::str::FromStr;
+
+/// This attribute macro generates the boilerplate required to call into the
+/// contract-specific logic from the entry-points to the Wasm module.
+///
+/// It should be added to the contract's init, handle, migrate and query implementations
+/// like this:
+/// ```
+/// # use cosmwasm_std::{
+/// #     Storage, Api, Querier, DepsMut, Deps, entry_point, Env, StdError, MessageInfo,
+/// #     InitResponse, HandleResponse, QueryResponse,
+/// # };
+/// #
+/// # type InitMsg = ();
+/// # type HandleMsg = ();
+/// # type QueryMsg = ();
+///
+/// #[entry_point]
+/// pub fn init(
+///     deps: DepsMut,
+///     env: Env,
+///     info: MessageInfo,
+///     msg: InitMsg,
+/// ) -> Result<InitResponse, StdError> {
+/// #   Ok(Default::default())
+/// }
+///
+/// #[entry_point]
+/// pub fn handle(
+///     deps: DepsMut,
+///     env: Env,
+///     info: MessageInfo,
+///     msg: HandleMsg,
+/// ) -> Result<HandleResponse, StdError> {
+/// #   Ok(Default::default())
+/// }
+///
+/// #[entry_point]
+/// pub fn query(
+///     deps: Deps,
+///     env: Env,
+///     msg: QueryMsg,
+/// ) -> Result<QueryResponse, StdError> {
+/// #   Ok(Default::default())
+/// }
+/// ```
+///
+/// where `InitMsg`, `HandleMsg`, and `QueryMsg` are contract defined
+/// types that implement `DeserializeOwned + JsonSchema`.
+///
+/// This is an alternative implementation of `cosmwasm_std::create_entry_points!(contract)`
+/// and `cosmwasm_std::create_entry_points_with_migration!(contract)`
+/// and should not be used together.
+#[proc_macro_attribute]
+pub fn entry_point(_attr: TokenStream, mut item: TokenStream) -> TokenStream {
+    let cloned = item.clone();
+    let function = parse_macro_input!(cloned as syn::ItemFn);
+    let name = function.sig.ident.to_string();
+    // The first argument is `deps`, the rest is region pointers
+    let args = function.sig.inputs.len() - 1;
+
+    // E.g. "ptr0: u32, ptr1: u32, ptr2: u32,"
+    let typed_ptrs = (0..args).fold(String::new(), |acc, i| format!("{}ptr{}: u32,", acc, i));
+    // E.g. "ptr0, ptr1, ptr2,"
+    let ptrs = (0..args).fold(String::new(), |acc, i| format!("{}ptr{},", acc, i));
+
+    let new_code = format!(
+        r##"
+        #[cfg(target_arch = "wasm32")]
+        mod __wasm_export_{name} {{ // new module to avoid conflict of function name
+            #[no_mangle]
+            extern "C" fn {name}({typed_ptrs}) -> u32 {{
+                cosmwasm_std::do_{name}(&super::{name}, {ptrs})
```

Comment:
Looking at https://github.com/CosmWasm/cosmwasm/blob/b38e79083c6f0d9e9ec8490ca1e97158f871b34f/packages/std/src/exports.rs#L96-L115 I wonder if this do_{name} code could be auto-generated.

We have the concrete type info from the function we are decorating, the rest is mostly boilerplate.

(I would do that as a separate PR however)

Reply:
Jupp, much to explore from here.

I also though about Multi-value regions in #602, such that we can have always have one pointer input and one pointer output.

Reply:
> I also though about Multi-value regions in #602, such that we can have always have one pointer input and one pointer output.

I think that would actually make the auto-generation harder, as you are accepting a variable number of input slices and then mapping them to a fixed number of typed variables. 

But yeah, plenty of room to explore here :)